### ISSUE TITLE
associativity of existing operations

### ISSUE THREAD

**Issue by DanGrayson** [UniMath/UniMath#187]
It turns out nicely that stnsum, stnprod, and my product operation in a monoid
over a finite sequence are all left associative, and that the parser is left
associative for ```+``` and ```*```.  But, Nat.mul defines the product
```m*n``` as a right associated sum of copies of n!  One result is that ```1*m```
does not reduce to m, and it's important for it to do so, because stnprod
actually returns not ```a*b*c*d``` but ```1*a*b*c*d``` -- the induction starts
with the empty product.  Similarly, stnsum starts its induction with an empty
sum, but luckily ```0+m``` reduces to m.

We can fix this (see code below), but then a lot of proofs that Vladimir and I
did for natural numbers have to be changed.

Another idea would be to change the code for stnsum and stnprod so they treat
sequences of length 1 as a special case, enabling stnprod to return
```a*b*c*d```, which is intuitively what people expect, anyway.  But then
theorems proving things about stnsum and stnprod by induction will have to be
changed, to do a similar double induction.

Discussion?

Here is demo code.

```
(* confirm and repair some aspects of operations on [nat] *)

Goal ∀ n, 0+n = n.
Proof. reflexivity. Defined.

Goal ∀ n, n+0 = n.
Proof. try reflexivity. Abort.

Goal ∀ n, n*0 = 0.
Proof. try reflexivity. Abort.

Goal ∀ n, 0*n = 0.
Proof. reflexivity. Defined.

Goal ∀ n, n*1 = n.
Proof. try reflexivity. Abort.

Goal ∀ n, 1*n = n.
Proof. try reflexivity. Abort.

(* That failure is not good, and Nat.mul defines multiplication as a right-associative sum,
   whereas the parser treats "+" and "*" as left associative.  
   We redefine multiplication in nat here. *)

Fixpoint mul n m := match n with | 0 => 0 | S p => p * m + m end
                    where "n * m" := (mul n m) : nat_scope.

(* make sure we didn't spoil anything else *)

Goal ∀ n, n*0 = 0.
Proof. try reflexivity. Abort.

Goal ∀ n, 0*n = 0.
Proof. reflexivity. Defined.

Goal ∀ n, n*1 = n.
Proof. try reflexivity. Abort.

Goal ∀ n, 1*n = n.
Proof. reflexivity. Defined.
```

And here is Coq's definition:
```
Fixpoint mul n m :=
  match n with
  | 0 => 0
  | S p => m + p * m
  end

where "n * m" := (mul n m) : nat_scope.
```

**Comment by benediktahrens** [comment]
I have not needed yet the arithmetic operations on nat, nor the std 
finite sets, so it is a bit difficult for me to comment.

I have not understood whether your "Another idea" is an alternative or 
an option in addition to redefining multiplication.

How much overhead does it bring to have 1*a*b*c*d instead of a*b*c*d ?
Is it mostly for cosmetic reasons that you would prefer the latter, or 
is it difficult to work with?

**Comment by DanGrayson** [comment]
No, I don't think it's difficult, but it is surprising to have basic operations defined
in non-analogous ways, not just to us, but also to future users of UniMath, so
it's not just cosmetic.

Yes, the other idea was meant as an alternative, but I don't like it much.

**Comment by vladimirias** [comment]
Redefining mul seems reasonable. It is indeed not a good definition when neither 1*n nor n*1 definitionally equal to n. I think there work of reproving results about multiplication will be minimal since not many of these results use the definition directly.

**Comment by DanGrayson** [comment]
Okay, I'll try it.  I've assigned the issue to me.

**Comment by DanGrayson** [comment]
Resolved in #189
